Documentation/Coherency.md

```diff
@@ -0,0 +1,86 @@
+
+Pros:
+- Simple.
+- Already have this mechanism.
+
+Cons:
+- We continue to not ship bits that match official builds, except on releases.
+- We have to deal with the problem of not getting all packages for servicing releases:
+	-  Issues https://github.com/dotnet/source-build/issues/210 and https://github.com/dotnet/source-build/issues/923 are our earlier discussions.
+	-  If we turn on BuildAllConfigurations and BuildAllPackages, we will ship bits that do not exist except on RHEL machines.  If someone shares a project but not the packages directory, restoring these packages that do not officially exist will fail.
+	-  In the worst case, these packages that we force to build could be known-bad code (that is okay normally because that package is not shipping).
+	-  If we don't turn on BuildAllConfigurations and BuildAllPackages, we will continue to have a large number of prebuilt packages.
+
+## Build multiple versions of each repo
+This would mean we build CLI with its expected version of CoreFX, and core-setup with its own expected version of CoreFX.
+
+Pros:
+- We match the official bits very closely.
+- We don't have to deal with servicing considerations.
+
+Cons:
+- This explodes rather quickly and would have extremely long build times.
+- We have to be careful about cross-contaminating package feeds during the builds.
+
+## Store and download packages in later builds
+During each build, we would save source-built packages off in an Azure account or NuGet feed.  During later builds, we would restore these packages and use them in builds later in the dependency tree.  In this scenario, the packages are more akin to build intermediate files and not final products.
+
+Pros:
+- Improved build times.
+	- Fringe benefit: incremental builds would probably finally work.
+- Still lets us eliminate prebuilt packages (all packages stored are source-built).
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the production dev and daily builds.
+- Need to carry required packages in a tarball or to an N+1 build (could be nupkg, tar.gz, RPM, deb).
+- Probably does not satisfy requirements for no prebuilt binaries.
+
+## Publish official source-built packages
+We would publish actual source-built RPM/deb/etc packages during the build for use during future builds.  This is like the store-and-download solution, but allows for distro maintainers to use our produced packages instead of having to rebuild every repo-sha included in a release, which be an additional set of all repos for each servicing release.
```

Comment:
> We would publish actual source-built RPM/deb/etc packages during the build for use during future builds. [...] allows for distro maintainers to use our produced packages

I don't see why distro maintainers would be able to use the Microsoft-built bits in this situation. Wouldn't it be the same as putting them in the tarball as prebuilts, policy-wise?

My suggestion was that when the distro build runs, one of its outputs is an RPM/Deb of the artifacts produced by that distro build that could be used by future servicing builds. The distro maintainer would carry that forward into the next build to preserve building from source.

(With "we" meaning "the source-build infra" and "publish" meaning "make available as a distro build artifact", this section might already mean this, but it isn't completely clear to me. 😄)

Reply:
Oh, I see.  Yeah, that makes more sense to me.  Fixed.